driver-core/src/main/com/mongodb/internal/dns/JndiDnsClient.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.dns;
+
+import com.mongodb.MongoClientException;
+import com.mongodb.spi.dns.DnsClient;
+import com.mongodb.spi.dns.DnsException;
+import com.mongodb.spi.dns.DnsWithResponseCodeException;
+
+import javax.naming.Context;
+import javax.naming.NameNotFoundException;
+import javax.naming.NamingEnumeration;
+import javax.naming.NamingException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.InitialDirContext;
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.List;
+
+public class JndiDnsClient implements DnsClient {
+    private final InitialDirContext dirContext;
+
+    public JndiDnsClient() {
+        dirContext = createDnsDirContext();
+    }
+
+    @Override
+    public List<String> getAttributeValues(final String name, final String type) throws DnsException {
+        try {
+            Attribute attribute = dirContext.getAttributes(name, new String[]{type}).get(type);
+            if (attribute == null) {
+                return null;
+            }
+            List<String> attributeValues = new ArrayList<>();
+            NamingEnumeration<?> namingEnumeration = attribute.getAll();
+            while (namingEnumeration.hasMore()) {
+                attributeValues.add((String) namingEnumeration.next());
+            }
+            return attributeValues;
+        } catch (NameNotFoundException e) {
+            throw new DnsWithResponseCodeException(e.getMessage(), 3, e);
+        } catch (NamingException e) {
+            throw new DnsException(e.getMessage(), e);
+        } finally {
+            try {
+                dirContext.close();
```

Comment:
Done

Reply:
> I think it means that closing an InitialDirContext does not render it unusable, since we always look up both TXT and SRV records and therefore it has to be used at least twice.

My guess is that it's because we call the methods on different instances: `DefaultDnsResolver.resolveAdditionalQueryParametersFromTxtRecords` is called only from the constructor of `ConnectionString` always on a new instance of `DefaultDnsResolver`.